src/H5FDsubfiling/H5subfiling_common.c

```diff
@@ -901,85 +815,84 @@ init_subfiling(const char *base_filename, uint64_t file_id, H5FD_subfiling_param
+        HMPI_GOTO_ERROR(FAIL, "MPI_Comm_set_errhandler failed", mpi_code);
 #else
 #error "MPI-3 required for MPI_Comm_split_type"
 #endif
 
+    /* Use the file's index to create a new subfiling context ID */
+    if ((context_id = H5FD__subfiling_new_object_id(SF_CONTEXT)) < 0)
+        HGOTO_ERROR(H5E_VFL, H5E_CANTGET, FAIL, "couldn't create new subfiling context ID");
+
     /*
-     * Setup the application topology information, including the computed
+     * Initialize the application topology information, including the computed
      * number and distribution map of the set of I/O concentrators
      */
-    if (init_app_topology(context_id, subfiling_config, comm, node_comm, &app_topology) < 0)
-        H5_SUBFILING_GOTO_ERROR(H5E_VFL, H5E_CANTINIT, FAIL, "couldn't initialize application topology");
-
-    new_context->sf_context_id = context_id;
+    if (H5FD__subfiling_init_app_topology(context_id, subfiling_config, comm, node_comm, &app_topology) < 0)
+        HGOTO_ERROR(H5E_VFL, H5E_CANTINIT, FAIL, "couldn't initialize application topology");
 
-    if (init_subfiling_context(new_context, base_filename, file_id, subfiling_config, app_topology, comm) < 0)
-        H5_SUBFILING_GOTO_ERROR(H5E_VFL, H5E_CANTINIT, FAIL,
-                                "couldn't initialize subfiling application topology object");
-    new_context->sf_node_comm = node_comm;
+    /* Initialize the subfiling context */
+    if (H5FD__subfiling_init_context(context_id, prefix_env, base_filename, file_id, subfiling_config,
+                                     app_topology, comm, node_comm, &context) < 0)
+        HGOTO_ERROR(H5E_VFL, H5E_CANTINIT, FAIL, "couldn't initialize subfiling context");
+    assert(context);
+    node_comm = MPI_COMM_NULL; /* New context takes ownership of node_comm */
 
     *context_id_out = context_id;
 
 done:
-    if (config_file && (EOF == fclose(config_file)))
-        H5_SUBFILING_DONE_ERROR(H5E_FILE, H5E_CANTCLOSEFILE, FAIL,
-                                "couldn't close subfiling configuration file");
+    if (config_file && EOF == fclose(config_file))
+        HDONE_ERROR(H5E_VFL, H5E_CANTCLOSEFILE, FAIL, "couldn't close subfiling configuration file");
 
     H5MM_free(file_basename);
     H5MM_free(subfile_dir);
 
     if (ret_value < 0) {
-        if (app_topology) {
-            if (H5_free_subfiling_topology(app_topology) < 0)
-                H5_SUBFILING_DONE_ERROR(H5E_VFL, H5E_CANTFREE, FAIL, "couldn't free subfiling topology");
-        }
+        if (app_topology)
+            if (H5FD__subfiling_free_topology(app_topology) < 0)
+                HDONE_ERROR(H5E_VFL, H5E_CANTFREE, FAIL, "couldn't free subfiling topology");
 
-        if (H5_mpi_comm_free(&node_comm) < 0)
-            H5_SUBFILING_DONE_ERROR(H5E_VFL, H5E_CANTFREE, FAIL, "couldn't free MPI communicator");
+        if (node_comm != MPI_COMM_NULL)
+            if (H5_mpi_comm_free(&node_comm) < 0)
```

Comment:
Not a big deal at all, but `H5_mpi_comm_free` already checks for `MPI_COMM_NULL` and `MPI_COMM_WORLD`, which is why the code previously just called the function for simplicity.

Reply:
Ah, missed that, thanks.  Fixed now.